toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1073 @@
+    // Since our watch descriptor is an int, we use nsUint32HashKey to compute the
+    // hashing key. We need 2 indexes in order to quickly look up the
+    // changed resource in the Worker Thread.
+    nsClassHashtable<nsStringHashKey, WatchedResourceDescriptor>
+    mWatchedResourcesByPath;
+    nsDataHashtable<nsUint32HashKey, WatchedResourceDescriptor*>
+    mWatchedResourcesByHandle;
+
+    // The same callback can be associated to multiple watches so we need to keep
+    // them alive as long as there is a watch using them. We create two hashtables
+    // to map directory names to lists of nsMainThreadPtr<callbacks>.
+    nsClassHashtable<nsStringHashKey, ChangeCallbackArray> mChangeCallbacksTable;
+    nsClassHashtable<nsStringHashKey, ErrorCallbackArray> mErrorCallbacksTable;
+
+    // Other methods need to know that a shutdown is in progress.
+    bool mShuttingDown;
+
+    // Main inotify file descriptor (initialized in NativeFileWatcher's Init())
+    int mInotifyFileDescriptor;
+
+    // Queue for the events read from inotify file descriptor
+    std::queue<inotify_event*> mInotifyEventQueue;
+
+    nsresult RunInternal();
+
+    nsresult DispatchChangeCallbacks(
+            WatchedResourceDescriptor* aResourceDescriptor,
+            const nsAString& aChangedResource);
+
+    nsresult ReportChange(
+            const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+            const nsAString& aChangedResource);
+
+    nsresult DispatchErrorCallbacks(
+            WatchedResourceDescriptor* aResourceDescriptor,
+            nsresult anError,
+            long anOSError);
+
+    nsresult ReportError(
+            const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError,
+            nsresult anError,
+            long anOSError);
+
+    nsresult ReportSuccess(const nsMainThreadPtrHandle<
+                           nsINativeFileWatcherSuccessCallback>& aOnSuccess,
+                           const nsAString& aResourcePath);
+
+    void AppendCallbacksToHashtables(
+            const nsAString& aPath,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError);
+
+    void RemoveCallbacksFromHashtables(
+            const nsAString& aPath,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError);
+
+    nsresult MakeResourcePath(WatchedResourceDescriptor* changedDescriptor,
+                              const nsAString& resourceName,
+                              nsAString& nativeResourcePath);
+};
+
+/**
+ * The watching thread logic.
+ *
+ * @return NS_OK if the watcher loop must be rescheduled, a failure code
+ *         if it must not.
+ */
+nsresult
+NativeFileWatcherIOTask::RunInternal()
+{
+    int mInotifyReadTimeout;
```

Comment:
If it's a local variable, don't prefix it with `m`. The entire point of this `m`/`g`/`s` prefix is to make the difference between local variables, member variables, global variables and static variables :)

Also, what is this?